src/components/VideoPlayer/index.js

```diff
@@ -0,0 +1,166 @@
+/* eslint-disable no-underscore-dangle */
+import {ResizeMode, Video} from 'expo-av';
+import PropTypes from 'prop-types';
+import React, {useCallback, useEffect, useRef, useState} from 'react';
+import {View} from 'react-native';
+import FullScreenLoadingIndicator from '@components/FullscreenLoadingIndicator';
+import {usePlaybackContext} from '@components/VideoPlayerContexts/PlaybackContext';
+import useWindowDimensions from '@hooks/useWindowDimensions';
+import addEncryptedAuthTokenToURL from '@libs/addEncryptedAuthTokenToURL';
+import stylePropTypes from '@styles/stylePropTypes';
+import styles from '@styles/styles';
+import VideoPlayerControls from './VideoPlayerControls';
+
+const propTypes = {
+    url: PropTypes.string.isRequired,
+
+    shouldPlay: PropTypes.bool,
+
+    onVideoLoaded: PropTypes.func,
+
+    resizeMode: PropTypes.string,
+
+    isLooping: PropTypes.bool,
+
+    style: stylePropTypes,
+
+    videoStyle: stylePropTypes,
+
+    shouldUseSharedVideoElement: PropTypes.bool,
+};
+
+const defaultProps = {
+    shouldPlay: false,
+    onVideoLoaded: () => {},
+    resizeMode: ResizeMode.CONTAIN,
+    isLooping: false,
+    style: [styles.w100, styles.h100],
+    videoStyle: [styles.w100, styles.h100],
+    shouldUseSharedVideoElement: false,
+};
+
+function VideoPlayer({url, resizeMode, shouldPlay, onVideoLoaded, isLooping, style, videoStyle, shouldUseSharedVideoElement}) {
+    const {isSmallScreenWidth} = useWindowDimensions();
+    const {currentlyPlayingURL, updateSharedElements, sharedElement, originalParent, updateCurrentVideoPlayerRef, currentVideoPlayerRef, updateIsPlaying} = usePlaybackContext();
+    const [duration, setDuration] = useState(0);
+    const [position, setPosition] = useState(0);
+    const [isLoading, setIsLoading] = useState(true);
+    const videoPlayerRef = useRef(null);
+    const videoPlayerElementParentRef = useRef(null);
+    const videoPlayerElementRef = useRef(null);
+    const sharedVideoPlayerParentRef = useRef(null);
+    const sourceURLWithAuth = addEncryptedAuthTokenToURL(url);
+
+    const onReadyForDisplay = useCallback(
+        (e) => {
+            onVideoLoaded(e);
+        },
+        [onVideoLoaded],
+    );
+
+    const onPlaybackStatusUpdate = useCallback(
+        (e) => {
+            updateIsPlaying(e.isPlaying);
+            setIsLoading(Number.isNaN(e.durationMillis)); // when video is ready to display duration is not NaN
+            setDuration(e.durationMillis || 0);
+            setPosition(e.positionMillis || 0);
+        },
+        [updateIsPlaying],
+    );
+
+    const bindFunctions = useCallback(() => {
+        currentVideoPlayerRef.current._onPlaybackStatusUpdate = onPlaybackStatusUpdate;
```

Comment:
In the documentation, I found [a bit different approach](https://docs.expo.dev/versions/latest/sdk/av/#example-setonplaybackstatusupdate) for setting this. Did this approach not work here?

Reply:
The purpose of `bindFunctions` is to synchronize in-chat video player states with the shared video inside attachment modal. On wider screens to optimize loadings when user is switching between chat an the modal we share expo video component. We want to make modal's <VideoPlayer> component have the same state as in-chat player